### DISCUSSION TITLE
STM32G4 Baud Rate

### DISCUSSION THREAD

**Discussion by kyleb-ici** [stm32duino/Arduino_Core_STM32#1509] [Q&A]
I'm using the STM32G431KB Nucleo 32 board with stm32duino, and I'm trying to get faster baud rates working. I am able to get 1Mbps working, but anything higher does not work. I've tried 2, 4, 8, and 12 Mbps.

I tried using STM32CubeMX to figure out the exact baud rate that I should use given the clock parameters specified in this chip's variant generic_clock.c file. But using the values that CubeMX specifies doesn't seem to make it work.

The STLink VCP says it can handle up to 16Mbps, so that shouldn't be the limiting factor.

I'm having a similar issue on the H743 Nucleo board, where I can only go up to 4Mbps.

**Comment by fpistm** [comment]
By default the Nucleo G431KB uses the HSI as default clock.
The clock config is not in the generic_clock.c but in the https://github.com/stm32duino/Arduino_Core_STM32/blob/master/variants/STM32G4xx/G431K(6-8-B)(T-U)_G441KB(T-U)/variant_NUCLEO_G431KB.cpp#L64

Several things can interfere with high speed: clock precision, wiring length, application code,....
Moreover to reach HS DMA is often involved which is not the case in this core.

**Reply by kyleb-ici** [reply]
That's helpful info, thanks. It looks like the clock config is the same as the generic clock file though, which also matches the clock config in CubeMX, as far as I can tell:
![image](https://user-images.githubusercontent.com/83553990/135311733-021dee9a-bd0f-4b39-914d-8e04d0a931fb.png)

I suppose the buffer could be filling up before I'm able to read from it. I did another test and I am able to send from the Nucleo to my laptop. Just not the other way around.

I might need to switch to Mbed then, because it supports buffered serial connections. Even at lower speeds that currently work, that could stop working as my code gets more complex if there's only one byte of buffer.

**Reply by kyleb-ici** [reply]
@fpistm  Actually it looks like I can change the buffer size: https://github.com/stm32duino/wiki/wiki/Custom-definitions#serial-rxtx-buffer-size

Is there a maximum allowed buffer size?

**Reply by fpistm** [reply]
Well index type used to access the buffer is an uint16_t so 65535 and of course the RAM size :)
